packages/eslint-plugin/src/rules/no-unnecessary-parameter-property-assignment.ts

```diff
@@ -0,0 +1,70 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+
+import * as util from '../util';
+
+export const messageId = 'noUnnecessaryParameterPropertyAssignment';
+
+export default util.createRule({
+  name: 'no-unnecessary-parameter-property-assignment',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Disallow unnecessary assignment of constructor property parameter to itself',
+      recommended: 'error',
+      requiresTypeChecking: false,
+    },
+    schema: [],
+    messages: {
+      [messageId]:
+        'Assignment has no effect as roperty parameter is already defined in the constructor.',
+    },
+  },
+  defaultOptions: [],
+  create(context) {
+    const parameterProperties = new Set<string>();
+    return {
+      'ClassDeclaration, ClassExpression'(): void {
+        parameterProperties.clear();
```

Comment:
[Bug] Clearing `parameterProperties` on each new class won't hold up for inner classes declared inside outer classes. Consider this unfortunate bit of code:

```ts
class First {
  constructor(public a: number) {
    class Second {
      constructor(public a: number) {
        this.a = a;
      }
    }

    new Second(0);

    this.a = a;
  }
}
```

...and also imagine if there are variations of that where either or both `a`s are renamed to `b` and/or not followed with an unnecessary parameter property assignment.

Perhaps keeping a stack of the current class would be helpful, similar to how `no-invalid-this` has a `thisIsValidStack: boolean[]`?